### DISCUSSION TITLE
The rectangle drawn by Graphics.drawRect() has faults

### DISCUSSION THREAD

**Discussion by shaucky** [airsdk/Adobe-Runtime-Support#2362] [Q&A]
From a personal project of mine. The rectangle drawn by Graphics.drawRect() has faults, but I didn't test out why. At first I thought it was because I called too many properties, so I optimized the code. But this bug still exists. The effect is as shown in the figure:

![faults drawn by Graphics.drawRect()](https://user-images.githubusercontent.com/119972730/208268437-0d82eb3b-8162-4309-a968-6fdf56861d26.png)

And the ActionScript goes here: 

	public class ViewPanel extends Panel {
		private const viewArea: Sprite = new Sprite();
		private var _backgroundColor: Number = 0xffffff;
		public function set backgroundColor(value: Number) {
			_backgroundColor = value;
			viewArea.graphics.clear();
			viewArea.graphics.beginFill(_backgroundColor);
			viewArea.graphics.lineStyle(1, 0x000000);
			viewArea.graphics.drawRect(0, 0, stage.stageWidth - 1, stage.stageHeight - 40 - 16);
			viewArea.graphics.endFill();
		}
		public function get backgroundColor(): Number {
			return _backgroundColor;
		}

		public function ViewPanel(parentFrame: Sprite) {
			super(parentFrame);
			title = "视口";
			tag = "view";
			((getChildAt(0) as Sprite).getChildAt(0) as TextField).text = window.title;
			window.width = 450;
			defaultWidth = 450;
			window.height = 300;
			defaultHeight = 300;
			stage.addEventListener(Event.RESIZE, stageResize);
			stage.addEventListener(Event.COMPLETE, contentLoaded);
		}
		private function resizeContent() {
			content.addChild(viewArea);
			viewArea.x = 0;
			viewArea.y = 16;
			viewArea.graphics.clear();
			viewArea.graphics.beginFill(_backgroundColor);
			viewArea.graphics.lineStyle(1, 0x000000);
			viewArea.graphics.drawRect(0, 0, stage.stageWidth - 1, stage.stageHeight - 40 - 16);
			viewArea.graphics.endFill();
		}
		private function stageResize(e: Event) {
			if (content) {
				resizeContent();
			}
		}
		private function contentLoaded(e: Event) {
			stage.removeEventListener(Event.COMPLETE, contentLoaded);
			resizeContent();
		}

	}

It looks like it fires after taking up a certain amount of memory. On my Windows the threshold is about 84MB.

Has anyone encountered a similar problem? How could this issue be triggered and is there a solution?

Thank you all!

**Comment by JohnBlackburne** [comment]
That looks a lot like something I've seen before, when you hold and drag the lower right corner of the window, and it leaves behind part drawn artefacts. The problem is it's firing `Event.RESIZE` events every frame, so maybe 60 times a second or more, and the renderer can't keep up, it gets interrupted and so doesn't finish drawing.

In my case though it wasn't a good test of the renderer, as dragging the corner of the window was not something that would be possible in the final app. On mobile you would expect a resize event once at startup, and often that's all, unless you allow for screen rotation which could send resize events but only separated by seconds.

So I added code to only handle resize events every second or so, effectively ignoring ones that arrived between:
```

		// ...
		stage.addEventListener(ResizeEvent.RESIZE, onResize);
		// ...

		private function onResize(e:ResizeEvent):void {
			mResize = e;
			e.stopImmediatePropagation();
		}

		private function ResizeCheck():void {

			if (iResizeTimer > 0) {
				iResizeTimer--;
			} else {
				if (mResize) {
					rStage.setTo(0, 0, mResize.width, mResize.height);

					// tell Starling
					Starling.current.viewPort = rStage;
					Starling.current.stage.stageWidth = rStage.width;
					Starling.current.stage.stageHeight = rStage.height;

					// tell stage
					Main.mMain.Resize(rStage);
			
					// update game objects
					OnResize();
					gDMgr.OnResize();
					iResizeTimer = 50;
					mResize = null;
				}
			}
		}
```

Even if you have an app users can resize, on desktop. doing something like I did I think makes sense. Or just ignore resize events until they stop arriving, then half a second after they stop use the last one you received.

**Reply by shaucky** [reply]
Thanks for your answer and your ideas, but my problem is still not solved.

I tried your suggestion to add an `uint` to count time, but doesn't matter. Maybe it's because i instantiate many `NativeWindow` and evey stage of them has event listener to draw graphics with `Event.RESIZE`?

Now i tried to add an event listener with `Event.EXIT_FRAME` to draw the rectangle, but it's still has faults.

I've test just now. Maybe the faults are not cased by these ActionScript. It's about other ActionScript that controls the `NativeWindow` cases. Let me think about it again…

**Reply by ajwfrost** [reply]
Are you able to set up a reproduction of this without Flex? Your sample code derives from Panel which I assume is `mx.containers.Panel` or `spark.components.Panel`.. With all the complexity that Flex adds around resizing and layouts, it would be good to have a pure AS3/AIR test case and see if this can be reproduced without the other component drawing.

**Reply by shaucky** [reply]
Sorry for i didn't describe clearly. The `Panel` is a custom class. The whole related ActionScript program seems to be complicated.
Simply, I have a `NativeWindow` as main window, and many `NativeWindow` owned by main window. The `stage` of main window has an event listener with `Event.RESIZE`, the handler will change size and position of every owned window by main window(move and resize with main window). The sample is one of the owned window's `stage`'s display object. Then as you see, the display object add an event listener for `Event.RESIZE` with its stage, to draw the rectangle.

**Comment by JohnBlackburne** [comment]
You should be able to add code to control when it responds to `Event.RESIZE` events. I've added the code I used to the above comment, where it registered the event but the handler `onResize` just stores  the event, for `ResizeCheck ` to periodically grab it from.

The other option is simply disable dynamic resizing. On mobile you don't have to worry about it normally. On PC it's common in my experience for games to run at a fixed number of resolutions, corresponding to common screen resolutions. Changing resolution only happens rarely, often requires assets to be reloaded, sometimes needs the game to restart.

**Comment by 2jfw** [comment]
```
	private function resizeContent() {
		content.addChild(viewArea);
		viewArea.x = 0;
		viewArea.y = 16;
```

Small hint: adding the child all the time on resize does not look right to me.
Also setting x and y all the time is overhead - try to only do that once

**Reply by 2jfw** [reply]
I would also recommend you to set the `stage.frameRate` to **60**

**Reply by shaucky** [reply]
Thanks, the previous part of the code is somewhat redundant and should indeed be moved to a place where it is executed only once.
I'm always used to setting the `stage.frameRate` to 24 or 25, and have no experience with the relationship between frame rate and graphics. Does frame rate affect API drawing?

**Reply by JohnBlackburne** [reply]
I think frame rate determines the rate you get `Event.ENTER_FRAME` events. `ResizeEvent.RESIZE` events though I think are generated additionally and independently, whenever a window's dimensions change. Which could be never, only occasionally, or continuously if you manually resize the window.

Hence how doing the latter can cause it to try and draw far more often than normal. Setting the frame rate low might be part of the problem – if your code is designed to redraw everything in 1/24 of a second then getting an _additional_ 60+ `ResizeEvent.RESIZE` events a second might be far too much for it to deal with.

**Reply by shaucky** [reply]
I've tried just now. Setting a high frame rate (such as 60) didn't solve the problem.

It's unknown for me that the logic of display object drawing. What i mean is, um, as we know, AVM2 is single-threaded, so what i understood is if the prev handler didn't complete, the next handler won't be executed. If my understanding is correct, the renderer shouldn't gets interrupted by new `Event`, so the problem shouldn't be happened by this.

**Reply by ajwfrost** [reply]
Actually, AIR isn't single-threaded; the ActionScript processing is, but the rendering is sliced and different parts of the stage are drawn in parallel. But the synchronisation is fairly straightforward so I don't think it should be possible for drawing APIs to be happening whilst the previous rendering is still going on, so you're essentially right in terms of everything being sequential. At least that's how it should work; when we have multiple windows, there may be some added complexities between the runtime vs the OS windowing system..!

This is happening on Windows then? You commented:
> It looks like it fires after taking up a certain amount of memory. On my Windows the threshold is about 84MB.

So does that mean, when you first start up the app, it all behaves fine, but if you've got something else that eats up memory then at some point this glitching effect starts happening?

If you're able to create a sample app so that we can reproduce this here, we can try to see what's happening..?

**Comment by shaucky** [comment]
[sandbox.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/10284549/sandbox.zip)

I deleted the irrelevant files in this personal project, left the parts related to this problem, and put them in this ZIP file. I think it's so complex for me to recreate a sample, and feel so sorry. Wishing this would not so complex for you. Thanks.

This part of the program is used by selecting 窗口(W) > 视口 menu item to activate the owned `NativeWindow`, and drag the owned window to the four edges of the main window, then drag or resize the main window, about half a minute, the problem will happen.

The problem is not only happens on Windows, but also Macintosh. I mentioned Windows in my comment because I only call the task manager of Windows to check the memory usage. Oh, by the way, `flash.system.System.privateMemory` seems incorrect on Windows, it has a big gap with task manager and the same program on Macintosh.

**Reply by shaucky** [reply]
I also know my code structure is so terrible, so ashamed…

**Reply by ajwfrost** [reply]
:-)

So I'm wondering if there's some conflicts going on between the resize handling in the base class (`WindowPanel` class) vs the derived `ViewPanel` one, and/or the use of `NativeWindow.width` vs `Stage.stageWidth` and the like.. 

The actual effect of this is intriguing: once there's a glitch on there, it stays and if you just resize the window a tiny bit it starts to move towards being a rectangle heading towards the bottom-right..

I'll ask someone to have a quick look at rebuilding this app and trying out a few things, plus we can add hooks into the drawing APIs to see what's happening there -> but the fact it's keeping the mis-information around between frames makes me think it's more likely to do with some rounding issues that are going on with internal render buffers rather than the drawing API itself..

thanks

**Reply by shaucky** [reply]
My thought is, the program as a desktop Application, should have some code to control windows, this is why `WindowPanel` use properties of `NativeWindow`. The derived class actually controls specific display object, so this is why i choose properties of `Stage`. For more, my thoughts on these two situations is because of a small detail between `Event.RESIZE` and `NativeWindowBoundsEvent.RESIZE`. The latter fires rate is lower than the former. If i use the latter to handle display object, it's unsmooth and unnatural.

I reviewed this part ActionScript carefully today. The using about properties of `NativeWindow` and `Stage` are indeed disordered… At first i used `NativeWindow` all the time, but then i find out that for display object, the effect that use properties of `Stage` is more smooth than use properties of `NativeWindow`. So in the code i wrote later, i implemented the latter idea, but for old code i didn't modified.

**Reply by shaucky** [reply]
Excuse me, do you have any idea about this problem?

My project I wrote for a competition is about to be submitted… (Although it may well be too late to wait for fixing the problem) However i wish the problem would be fixed.

Later in development, I realized that this faults seemed to be at the top of the content, maybe it's at the end of AIR drawing…? (Just a conjecture without foundation)

@ajwfrost Thank you.

**Reply by shaucky** [reply]
@ajwfrost Did this bug be tested at the time and what was the problem? Today i saw another guy in a chat group met an issue looks like these. And AIR 51.0 has not fixed this yet.
 
Thanks.

**Reply by ajwfrost** [reply]
Hi @shaucky - I remember us looking into this one, I believe we could reproduce it but weren't able to work out what was actually going wrong. I don't recall if we could reproduce this without using Flex though - it's very possible that the Flex rendering code is requesting some drawing with incorrectly rounded values which would then be the reason that the problems are then staying on the screen despite full refreshes of the frame buffer.

One thing we've talked about internally for quite a while is a way to break down the display list so it can be inspected from the outside (e.g. via Scout) which would then give us more insights as to where this kind of thing is going wrong. If we can get that mechanism set up, it should help narrow this one down..

thanks

**Reply by MalacTheLittle** [reply]
Not sure if it's related to this issue, but looks similar and maybe gives you a hint how to resolve it... 
Defining `graphics.lineStyle.scaleMode = "none"` in combination with `graphics.lineStyle.joints = "miter"` results with same glitch when draging object. All other combinations works fine. If window is resized it renders to desired result. Here's simple test case (grab shape and drag it around):

![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/56675407/2f471868-0053-4e27-922c-72645790fafe)
and it should look like this (which is achieved with manual window resize):
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/56675407/85e72ea5-33b3-4987-9a39-c845ad5686c5)


```
import flash.display.Shape;
import flash.events.MouseEvent;
import flash.display.MovieClip;

var mc:MovieClip = new MovieClip();
addChild(mc);
var sh:Shape = new Shape();
sh.graphics.beginFill(0xFFFFFF);
sh.graphics.lineStyle(20, 0xFF0000, 0.5, false, "none", "none", "miter");
sh.graphics.drawRect(0,0,100,100);
sh.graphics.endFill();
mc.addChild(sh);

mc.addEventListener(MouseEvent.MOUSE_DOWN, onMDown);
function onMDown(event:MouseEvent):void{
	stage.addEventListener(MouseEvent.MOUSE_UP, onMUp);
	mc.startDrag();
}
function onMUp(event:MouseEvent):void{
	stage.removeEventListener(MouseEvent.MOUSE_UP, onMUp);
	stopDrag();
}
```
Obviously, outter part of stroke that exceeds shape size isn't considered in stage redraw when scaleMode is set to "none".
Tested on Windows with AIR 50.2.4.2 and Adobe Flash CC
It works fine on Android.

**Reply by shaucky** [reply]
Thanks for your contributing some ideas and cases. I think these two should not be the same issue. Anyway, I no longer expect to solve this problem that almost no one has encountered (the guy i mentioned before found his issue is caused of improper use of masks).
 
At first, I thought it was an API related issue with `Graphics`, but with further development, i found that this issue may not be related to the drawing of `Graphics`. It looks more like it's related to the entire Flash canvas (I'm not sure if this describe is appropriate).
 
My application later added some pixel complex bitmaps to the drawn content. Guess what? Those cracks which has same color with stage, display on top of these bitmaps, or rather, there are cracks on the Flash canvas itself.
 
Anyway anyway… I don't expect this problem to be solved anymore. It's been a long time and the related project has been abandoned by me due to some factors including this issue…
 
If anyone else is interested in the entire example with the issue, can also take a look at this repository: [https://github.com/shaucky/MapEditor](). But as my first attempt at complex development, it lacked annotations and documentation.
 
Thanks.

**Reply by ajwfrost** [reply]
Interesting one ... thanks @MalacTheLittle !
It's possible there may be a similar underlying cause for these, despite the fact they appear fairly different when you see them, and the original issue only happens after some memory usage has impacted things (really not sure why/how!).

Both of them appear to be caused by the Flash 'canvas' (I'd use the term 'frame buffer' but it's similar) not getting correct "dirty" areas that need to be cleaned up. We do have a way to display the "dirty" rectangles, which it uses to decide what to draw: potentially, the dirty rects aren't extended properly to cover these areas hence the old stuff being left behind until a full redraw was required (although with the original issue, I'm fairly sure we had tried updating to always set the whole display as 'dirty', but without it then helping...)

Anyway: this second case is so much easier to reproduce, we will fix that, and then see whether it has helped also with the original problem!

thanks

**Reply by ajwfrost** [reply]
@MalacTheLittle I'm not sure whether your screenshots were somehow set up with an additional, shaded area across the screen to create that two-tone pink/red effect in the static/redrawn version.. but, when I put that code into a SWF, I'm getting a white solid square with a wide pink border.. but if I move the white part of the square a small amount, the border doesn't change (it looks like you're moving a white square over a static pink square..).

When we enable the redraw regions, we can see it's only got half of the border as part of the redrawn area! So, it's not coping with the fact that the border is extending the shape beyond its normal location.

Seems very odd...! 'Bevel' and 'round' joints have the same extension of the shape, but without this problem. Which means it's likely that this one is trivial (and also less likely that it's the same root cause as the original issue, I guess)

thanks

**Reply by MalacTheLittle** [reply]
Ah, it's two toned because stage color is set to #333333, and stroke is red with 0.5 alpha, so it gets this shaded effect. I made it this way in order to emphasise that stroke is center aligned over the edge of a shape and actually "outer" part of a stroke is the problematic one.

> Seems very odd...! 'Bevel' and 'round' joints have the same extension of the shape, but without this problem. Which means it's likely that this one is trivial (and also less likely that it's the same root cause as the original issue, I guess)

It seems to me that "miter" joints with scaleMode "none" is the basic one and doesn't need any extra calculations, but for "round" and "bevel" joints and "normal" scaleMode it takes some extra recalculation for it's corners/weight.

**Reply by MalacTheLittle** [reply]
There's one more thing, if you set renderMode to "direct" there won't be dirt leftovers but outer part of a stroke is cutout.
